### PR TITLE
[onert] Introduce FullyConnectedLayer for training

### PR DESCRIPTION
**Pull request by ragmani**
This commit introduces FullyConnectedLayer for training.
  - Introduce `FullyConnectedBiasGrad` and its unit tests into cker
  - Introduce FullyConnectedLayer for training.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  {
+    case OperandType::FLOAT32:
+    {
+      assert(data_type == _grad_weights->data_type());
+      assert(data_type == _grad_bias->data_type());
+      backwardFloat32();
+      break;
+    }
+    default:
+      throw std::runtime_error{"train ConvolutionLayer: unsupported data type"};
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:105]:
```suggestion
      throw std::runtime_error{"train FullyConnectedLayer: unsupported data type"};
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+void FullyConnectedLayer::backward(uint32_t)
+{
+  const auto data_type = _deriv_output->data_type();
+  assert(data_type == _input->data_type());
+  switch (data_type)
+  {
+    case OperandType::FLOAT32:
+    {
+      assert(data_type == _grad_weights->data_type());
+      assert(data_type == _grad_bias->data_type());
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:100]: Can we move these asserts in front of the switch phrase?

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      break;
+    case ir::Activation::RELU:
+      ReLUGrad(getShape(_output), getBuffer<float>(_output), getShape(_deriv_output),
+               getBuffer<float>(_deriv_output), getShape(_act_deriv_output.get()),
+               getBuffer<float>(_act_deriv_output.get()));
+      act_grad = _act_deriv_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
+      break;
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:126]:
```suggestion
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      act_grad = _deriv_output;
+      break;
+    case ir::Activation::RELU:
+      ReLUGrad(getShape(_output), getBuffer<float>(_output), getShape(_deriv_output),
+               getBuffer<float>(_deriv_output), getShape(_act_deriv_output.get()),
+               getBuffer<float>(_act_deriv_output.get()));
+      act_grad = _act_deriv_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:125]:
```suggestion
      throw std::runtime_error("train FullyConnectedLayer: Unsupported activation type yet");
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+  nnfw::cker::TransposeParams transpose_param;
+  transpose_param.perm_count = 2;
+  transpose_param.perm[0] = 1;
+  transpose_param.perm[1] = 0;
+
+  // Initialize FullyConnectedParams
+  nnfw::cker::FullyConnectedParams op_params;
+  op_params.activation = nnfw::cker::FusedActivationFunctionType::kNone;
+
+  // Transpose and compute gradient for input
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:139]:
```suggestion
  // Transpose and compute gradient for input
  // dL/dx = dot(Incoming derivative, transpose W)
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  assert(transposed_weights->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_weights), getBuffer<float>(_weights),
+                        getShape(transposed_weights), getBuffer<float>(transposed_weights));
+
+  nnfw::cker::FullyConnected(op_params, getShape(act_grad), getBuffer<float>(act_grad),
+                             getShape(transposed_weights), getBuffer<float>(transposed_weights),
+                             getShape(nullptr), nullptr, getShape(_deriv_input),
+                             getBuffer<float>(_deriv_input));
+
+  // Transpose and compute gradient for weights
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:150]:
```suggestion
  // Transpose and compute gradient for weights
  // dL/dW = dot(Incoming derivative, transpose X)
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  auto transposed_input = _transposed_input.get();
+  assert(transposed_input->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_input), getBuffer<float>(_input),
+                        getShape(transposed_input), getBuffer<float>(transposed_input));
+
+  auto transposed_deriv_output = _transposed_deriv_output.get();
+  assert(transposed_deriv_output->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(act_grad), getBuffer<float>(act_grad),
+                        getShape(transposed_deriv_output),
+                        getBuffer<float>(transposed_deriv_output));
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:160]: (Q) Why do you transpose `deriv_output`?

refer: https://www.adityaagrawal.net/blog/deep_learning/bprop_fc

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                             getBuffer<float>(transposed_input), getShape(transposed_deriv_output),
+                             getBuffer<float>(transposed_deriv_output), getShape(nullptr), nullptr,
+                             getShape(_grad_weights), getBuffer<float>(_grad_weights));
+
+  // Compute gradient for bias
+  if (_bias)
+  {
+    assert(_grad_bias);
+    nnfw::cker::FullyConnectedBiasGrad(getShape(act_grad), getBuffer<float>(act_grad),
+                                       getShape(_grad_bias), getBuffer<float>(_grad_bias));
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/FullyConnectedLayer.cc:172]: (Q) What is benefit of implementing sum of incoming derivative in cker?

Comment (jyoungyun) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:172]: Oh, you might want to use Eigen function for this. Am I right?

File: compute/cker/src/train/FullyConnected.test.cc [hunk 9]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (jyoungyun) [inline compute/cker/src/train/FullyConnected.test.cc:1]: How about moving it in `test` directory?

`compute/cker/include/cker/train/test`

File: compute/cker/include/cker/train/operation/FullyConnected.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+#define __NNFW_CKER_TRAIN_OPERATION_FULLY_CONNECTED_H__
+
+#include "cker/eigen/Utils.h"
+#include "cker/Shape.h"
+
+namespace nnfw
+{
+namespace cker
+{
+
```
Comment (jyoungyun) [inline compute/cker/include/cker/train/operation/FullyConnected.h:27]: How about using `train` namespace?

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 2 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:100]: There are cases the data type of weights and bias is different from the data type of input, so it looks better to put them here.

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 5 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:139]:
```suggestion
  // Transpose and compute gradient for input
  // ∂L/∂X = matmul(Incoming gradient, transposed W)
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 6 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:150]:
```suggestion
  // Transpose and compute gradient for weights
  // ∂L/∂W = matmul(transposed X, transposed incomming gradient)
```

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 7 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:160]: Considering the shapes. transpose is required to use with fc kernel.

output {2, 3} = fc( input {2, 5}, weights {3, 5} )
∂L/∂X {2,5} = fc( incoming gradient {2, 3}, transposed weights {5, 3} )
∂L/∂W {3, 5} = transpose ( fc( transposed input {5, 2}, transposed incoming gradient {3, 2} ) )

I haven't checked the formula in detail. So I will check this implementation with the formula.

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 8 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:172]: Output is 2D and Bias is 1D, and `output.dim(1)` == `bias.dim(0)`. So The gradient of the bias is row-wise sum of the incoming gradients.

File: compute/cker/include/cker/train/operation/FullyConnected.h (hunk 10 above)
Comment (ragmani) [reply compute/cker/include/cker/train/operation/FullyConnected.h:27]: OK. I will add it.

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 8 above)
Comment (jyoungyun) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:172]: Um, my question is about implementation location. I wonder why you implemented the details in cker . However, when I implement loss, it seems appropriate to implement it in cker to use eigen function.

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 7 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc:160]: I confirmed the formula in detail. 

$fc (X, W^T) = matmul(X, W)$


output {2, 3} = fc( input {2, 5}, weights {3, 5} )$
∂L/∂X {2, 5} = fc( incoming gradient {2, 3}, transposed weights {5, 3} )
∂L/∂W {3, 5} = fc( transposed incoming gradient {3, 2},  transposed input {5, 2} )

output {2, 3} = matmul( input {2, 5}, transposed weights {5, 3} )
∂L/∂X {2, 5} = matmul( incoming gradient {2, 3}, weights {3, 5} )
∂L/∂W {3, 5} = matmul( transposed incoming gradient {3, 2},  input {2, 5} )

If $W$ is transposed weight, input is $X$, and output is $Z$,
$Z = matmul( X, W )$
$∂L/∂X = matmul( ∂L/∂Z, W^T )$
$∂L/∂W = matmul( (∂L/∂Z)^T , I )$

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 8 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc]: Ah..Sorry for my misunderstanding. 
If you want to use Eigen, you need to implement it in cker. As for other cases, I don't know exactly.
In fact, many of cker's code was imported from tensorflow. So, we imported the codes that had different formats in the cker. And we implemented layers in cpu backend to be able to call them in their formats.
I think it's okay to implement simple kernels wherever you feel comfortable.

File: runtime/onert/backend/train/ops/FullyConnectedLayer.cc (hunk 7 above)
Comment (jyoungyun) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc]: Could you let me know a link to the weight derivative information?

I found the following formula in your repo.

> #### Weights
> $\frac{\partial L}{\partial W} = (\frac{\partial L}{\partial Z})^T X^T$

But it's different from my formula. So I want to know the details.

> ### $\cfrac{\partial L}{\partial W} = \cfrac{\partial L}{\partial Z} X^{T}$

Comment (ragmani) [reply runtime/onert/backend/train/ops/FullyConnectedLayer.cc]: The formulas in the link are based on the forward $Z = WX + B$
My example formulas are based on the forward $Z = XW + B$ 

Matrix multiplication is not commutative, so we cannot use that formula in the link as it is.
For example, since $(AB)^T = B^TA^T$ and $AB = ((AB)^T)^T$, the above formula ($Z = WX + B$) is changed to $Z = (X^T W^T)^T + B$ so that the results of Z are the same.

If you think of $X$ as $W$ and $X$ as $W$, the formula I wrote and the formula in the link are the same.
$Z = matmul( W,  X)$
$∂L/∂X = matmul( (∂L/∂Z)^T , W )$
$∂L/∂W = matmul( ∂L/∂Z, X^T )$

**Review by jyoungyun — APPROVED** [review]
LGTM

**Review by hseok-oh — APPROVED** [review]
